### PR TITLE
UI

### PR DESCRIPTION
**Pull request by JereSalo**
## Summary
This PR is focused on the UI aspects of the game, such as drawing the cells on screen and the interaction of the user with the game.
The program can be run with `cargo run`, instructions are on screen when first opened.

### CODE REVIEW

**Comment by JereSalo** [comment]
Sorry that I didn't fill in the readme file. It will be the next thing I'll do ☺

File: src/main.rs [hunk 1]
```diff
@@ -1,23 +1,85 @@
...
+        clear_background(BLACK);
+
+        // Width and Height are defined on every tick cause screen can be resized.
+        let scr_w = screen_width();
+        let scr_h = screen_height();
+        let cell_w = scr_w / grid.cols() as f32;
+        let cell_h = scr_h / grid.rows() as f32;
+
+        // Draw Board
+        #[allow(clippy::needless_range_loop)]
```
Comment (iinaki) [inline src/main.rs]: I think you can remove this `#[allow]`, I tried running clippy without it and it threw no warnings

File: src/main.rs [hunk 2]
```diff
@@ -1,23 +1,85 @@
...
 async fn main() {
-    clear_background(BLACK);
-
-    let mut grid = Grid::new_from_cells(vec![
-        vec![Cell::Dead, Cell::Alive, Cell::Dead, Cell::Dead],
-        vec![Cell::Alive, Cell::Alive, Cell::Dead, Cell::Dead],
-        vec![Cell::Dead, Cell::Dead, Cell::Alive, Cell::Dead],
-        vec![Cell::Dead, Cell::Dead, Cell::Dead, Cell::Alive],
-    ]); // This is an example grid, for testing purposes only
+    let (rows, cols) = (50, 50);
```
Comment (iinaki) [inline src/main.rs]: `rows` and `cols` can be constants!

File: src/main.rs [hunk 3]
```diff
@@ -1,23 +1,85 @@
...
+            draw_line(cell_w * i as f32, 0.0, cell_w * i as f32, scr_h, 1.0, GRAY);
+        }
+
+        // Handle KeyPress: R (Random), C (Clear), Space (Play/Pause), I (Toggle Info)
+        match get_last_key_pressed() {
+            Some(KeyCode::R) => grid = Grid::new_random(rows, cols),
+            Some(KeyCode::C) => grid = Grid::new(rows, cols),
+            Some(KeyCode::Space) => paused = !paused,
+            Some(KeyCode::I) => show_info = !show_info,
+            Some(_) | None => {}
```
Comment (iinaki) [inline src/main.rs]: The options are great, I would just add one more, something like `X` to end the game. So you can handle the ending of the program graciously and the user doesn't have to run `ctrl + C`.

**Review by iinaki — COMMENTED** [review]
I'll approve after the 3 little comments I made but I have some suggestions that you can do for the MVP:

- A thing that I tend to do (because most codebases I saw do the same thing) is to not have much code in the `main.rs` module, just some initializations and something that looks like `ui.run();`. I think you can create a new struct `Ui` that handles the running of the program in its own separate module. 

- Also I would take advantage of the refactor you will have to do to not have one single function that does the running, create several like `draw_board` and `handle_key_press`.

File: src/main.rs (hunk 1 above)
Comment (JereSalo) [reply src/main.rs]: Done!

File: src/main.rs (hunk 2 above)
Comment (JereSalo) [reply src/main.rs]: Done!

File: src/main.rs (hunk 3 above)
Comment (JereSalo) [reply src/main.rs]: Great idea! Hadn't thought about that. I've added `Q` to end the game, also updated the info text accordingly. Thanks for the suggestion!

**Comment by JereSalo** [comment]
> I'll approve after the 3 little comments I made but I have some suggestions that you can do for the MVP:
> 
> * A thing that I tend to do (because most codebases I saw do the same thing) is to not have much code in the `main.rs` module, just some initializations and something that looks like `ui.run();`. I think you can create a new struct `Ui` that handles the running of the program in its own separate module.
> * Also I would take advantage of the refactor you will have to do to not have one single function that does the running, create several like `draw_board` and `handle_key_press`.

I changed these things in the latests commits I pushed. Let me know your thoughts about how I handled it! Treat it as an MVP. Thanks for the suggestions 😀

File: src/main.rs [hunk 4]
```diff
@@ -1,23 +1,13 @@
...
-        vec![Cell::Dead, Cell::Dead, Cell::Alive, Cell::Dead],
-        vec![Cell::Dead, Cell::Dead, Cell::Dead, Cell::Alive],
-    ]); // This is an example grid, for testing purposes only
+    let mut ui = UI::new();
 
     loop {
-        // Here I should draw the grid state on screen I guess.
-
-        grid.update();
+        ui.run();
```
Comment (iinaki) [inline src/main.rs]: Looks good, but I think it makes more sense to put the loop inside `ui.run()`, so main would end up with just the initialization of `Ui` and the call to run :)

**Review by iinaki — CHANGES_REQUESTED** [review]
Just left one comment on the code that's left to resolve, the other stuff looks so much better now!

One more thing I'd like you to add is to document the structs and each method following the `rustdoc` standard, if you've never heard of it you can read about it [here](https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html). Basically it is to use `///` syntax on comments, it generates special docs for each object and it looks pretty.

File: src/main.rs (hunk 4 above)
Comment (JereSalo) [reply src/main.rs]: Agree. Done!

**Comment by JereSalo** [comment]
> Just left one comment on the code that's left to resolve, the other stuff looks so much better now!
> 
> One more thing I'd like you to add is to document the structs and each method following the `rustdoc` standard, if you've never heard of it you can read about it [here](https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html). Basically it is to use `///` syntax on comments, it generates special docs for each object and it looks pretty.

I didn't realize those types of comments were special. I added brief descriptions for data types and methods. 
Thank you for the suggestion!

**Review by iinaki — APPROVED** [review]
Great work Jere 💯

**Pull request merged**